Added aliases support for custom models

dagobert opened the pull request:
Aliases are recognized for custom models. Association between model/alias and model path is generated on loader initialization.

This pull request is discussed here #2564.

bramstroker commented:
Thanks for the PR! Looks good on first sight. Will do a more thorough review coming days.
I have added a section in docs last week explaining how to run the tests: https://docs.powercalc.nl/contributing/integration-development/#running-the-tests.
When you have them running you can debug why a lot are failing now.
Also there needs to be a test added verifying the alias lookup correctly works which was added in this PR. 
Think you can extend the parametrize section of this test: tests.power_profile.loader.test_local.test_find_model, to add the scenarios for the aliases.

bramstroker commented:
And yes tests are probably failing because there is not initialize call being done in the unit tests for localloader.
So adding `await loader.initialize()` should probably fix the issue.

dagobert commented:
I have now the testing environment ready and started to check the tests.

bramstroker commented:
> I have now the testing environment ready and started to check the tests.
> 
> I found some things already that did not break the code but need to be addressed. I will get back to you when I worked through the tests and added some.

Cool, thanks for the update.

dagobert commented:
> > I have now the testing environment ready and started to check the tests.
> > I found some things already that did not break the code but need to be addressed. I will get back to you when I worked through the tests and added some.
> 
> Cool, thanks for the update.

I have new version which is much changed to the current. I have added many more tests. When I run only the test_local.py I get no error but when I run the complete test suite I get some. They are mainly related to localLoader.load_model(). You are using this in the tests to directly load a json model within the directory handed to the loader. This is not a clean solution and after researching your code it seems that you only use it in your testing facility and not in your normal code.

If this is correct you have programmed a function to allow a specific test to run and not created a test to check a function. I would suggest to change the tests.

bramstroker commented:
I'm not really sure what you mean. Would be best to commit the updated code than I'll have a look this weekend.

dagobert commented:
Ok. going back to the immediate things and talk about the stuff from above later.

This commit is more or less final. The checks failing are probably bacause I am saving only lower case manufacturer/model/aliases in the library I generate from the directory. Hence, if you query models you will get back all lower case. In the tests you expect mixed case.

Either you need to have mixed case and I need to change `local.py` or the tests need to reflect this change.

bramstroker commented:
Casing test are in place to support the case you requested here about the mixed casing:
https://github.com/bramstroker/homeassistant-powercalc/discussions/2564
So they are there to prevent regressions and ensure this case keeps working now and in the future.
Apparently the changes in this PR broke this scenario, so we should fix the code and not change the tests.
I will dive into it later if you are stuck.

dagobert commented:
> Casing test are in place to support the case you requested here about the mixed casing: #2564 So they are there to prevent regressions and ensure this case keeps working now and in the future. Apparently the changes in this PR broke this scenario, so we should fix the code and not change the tests. I will dive into it later if you are stuck.

Not sure. My request was to allow lower case and upper case not matter to the library. Meaning, it would be the same to request for model "HS300" and "hs300".  both would return a valid answer al long as there is a model folder "HS300"/"hs300" or an alias "HS300"/"hs300" respectively.

In my changes I remove all meaning of case-sensitivity by storing all lower case (but the path to the model.json files of cause) and converting all function requests that have manufacturer or model as a variable.

The single result to this inside the running code (not the test) is that eg. when a device is discovered by a local.py the the model is in lower case (see pic: "hmip-psm").

![grafik](https://github.com/user-attachments/assets/70079b95-d11c-4b87-86f7-263301e83d8f)

I dont think this anything to worry since this was already hapening with manufacturer anyways.

IMHO unless in your other code you require case sensitivity of the returned model - such as it might in `library.get_model_listing` everything thould be good and the tests can be changed.

bramstroker commented:
Did some first cleanups. There were some typing issues my editor was complaining about.
And the loading of library mapping was rather complicated I simplified this a bit. device_type is mandatory so we can just create the model_data dictionary one and attach to both model or alias mapping.
Also checks for `os.path.exists` are unnessecary as Python os code would be broken. This statements can never be true imo and we also cannot get test coverage on that.
See https://github.com/bramstroker/homeassistant-powercalc/pull/2610/commits/207b8564ff03ea160447db8df8b0905dfe063797
Have to go now, will have a further look later on.

bramstroker commented:
Did another commit:
https://github.com/bramstroker/homeassistant-powercalc/pull/2610/commits/24dc23da4233c89a38d8a6762168e5cd32b0377e

Will have look at the tests next.

bramstroker commented:
I have created new PR #2625, because I cannot figure out how to push to your forked repo without using the GUI.
Have added some more commits to fix the tests and casing issues.

dagobert commented:
> Did some first cleanups. There were some typing issues my editor was complaining about. And the loading of library mapping was rather complicated I simplified this a bit. device_type is mandatory so we can just create the model_data dictionary one and attach to both model or alias mapping. 

Yes much nicer :)

bramstroker commented:
I mistakenly removed your tests, and replaced with my local copy in my new PR.
Will revive all the tests you added and verify / fix them

dagobert commented:
You closed the PR. Now I cannot keep help changing stuff :(

bramstroker commented:
> You closed the PR. Now I cannot keep help changing stuff :(

Yes I know :-(. I have added you as a collaborator. Let's see if I can give you write access to the branch `dagobert-aliases-for-custom-models-genlib` on my repo.
